sdks/java/core/src/main/java/org/apache/beam/sdk/io/BigQueryIO.java

```diff
@@ -553,10 +596,353 @@ public boolean getValidate() {
+        BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+        TableReference tableToExtract;
+        JobService jobService = bqServices.getJobService(bqOptions);
+        if (!Strings.isNullOrEmpty(query)) {
+          // Execute query
+          String queryJobId = jobIdToken + "-query";
+          tableToExtract = JSON_FACTORY.fromString(jsonQueryTempTable, TableReference.class);
+          executeQuery(queryJobId, query, tableToExtract, flattenResults, jobService);
+        } else {
+          tableToExtract = JSON_FACTORY.fromString(jsonTable, TableReference.class);
+        }
+        String extractJobId = jobIdToken + "-extract";
+        List<String> tempFiles = executeExtract(
+            extractJobId, tableToExtract, options.getTempLocation(), jobService);
+
+        TableSchema tableSchema = bqServices.getTableService(bqOptions).getTable(
+            tableToExtract.getProjectId(),
+            tableToExtract.getDatasetId(),
+            tableToExtract.getTableId()).getSchema();
+
+        return createSources(tempFiles, tableSchema);
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        // No way to estimate bytes, so returning 0.
+        return 0;
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return false;
+      }
+
+      @Override
+      public BoundedReader<TableRow> createReader(PipelineOptions options) throws IOException {
+        BigQueryOptions bqOptions = options.as(BigQueryOptions.class);
+        if (!Strings.isNullOrEmpty(query)) {
+          return new BigQueryReader(this, bqServices.getReaderFromQuery(
+              bqOptions, query, bqOptions.getProject(), flattenResults));
+        } else {
+          TableReference tableRef = JSON_FACTORY.fromString(jsonTable, TableReference.class);
+          return new BigQueryReader(this, bqServices.getReaderFromTable(bqOptions, tableRef));
+        }
+      }
+
+      @Override
+      public void validate() {
+        // Do nothing, validation is done in BigQuery.Read.
+      }
+
+      @Override
+      public Coder<TableRow> getDefaultOutputCoder() {
+        return TableRowJsonCoder.of();
+      }
+
+      private static void executeQuery(
+          String jobId,
+          String query,
+          TableReference destinationTable,
+          boolean flattenResults,
+          JobService jobService) throws IOException, InterruptedException {
+        JobConfigurationQuery queryConfig = new JobConfigurationQuery();
+        queryConfig
+            .setQuery(query)
+            .setAllowLargeResults(true)
+            .setCreateDisposition("CREATE_IF_NEEDED")
+            .setDestinationTable(destinationTable)
+            .setFlattenResults(flattenResults)
+            .setPriority("BATCH")
+            .setWriteDisposition("WRITE_EMPTY");
```

Comment:
I believe you need to also set the `location` of the temp table, or else BigQuery will use a default. This caused trouble for for tables in EU -- https://github.com/GoogleCloudPlatform/DataflowJavaSDK/issues/86